Libs/MRML/Core/vtkMRMLSubjectHierarchyNode.cxx

```diff
@@ -3008,14 +3069,28 @@ vtkMRMLSubjectHierarchyNode* vtkMRMLSubjectHierarchyNode::GetSubjectHierarchyNod
     vtkMRMLSubjectHierarchyNode* currentShNode = vtkMRMLSubjectHierarchyNode::SafeDownCast(node);
     if (currentShNode)
       {
-      if (!firstShNode->MergeSubjectHierarchy(currentShNode))
+      // Prevent events from being invoked from both subject hierarchy nodes
+      firstShNode->Internal->EventsDisabled = true;
+
+      // Copy all unresolved items from the merged subject hierarchy node into first node
+      firstShNode->Internal->UnresolvedItems->DeepCopy(currentShNode->Internal->UnresolvedItems);
+
+      // Remove merged subject hierarchy node from the scene
+      scene->RemoveNode(currentShNode);
+
+      // Resolve all unresolved items
+      if (!firstShNode->Internal->ResolveUnresolvedItems())
         {
         //TODO: The returned node will probably be invalid, so it needs to be completely re-built
-        //      (e.g. by removing all the subject hierarchy nodes from the scene and when adding a
-        //      new one, calling addSupportedDataNodesToSubjectHierarchy in the plugin logic)
+        // (e.g. by removing all the subject hierarchy nodes from the scene and when adding a
+        // new one, calling addSupportedDataNodesToSubjectHierarchy in the plugin logic)
         vtkErrorWithObjectMacro(scene, "vtkMRMLSubjectHierarchyNode::GetSubjectHierarchyNode: Failed to merge subject hierarchy nodes");
         return firstShNode;
         }
+
+      // Trigger re-building the trees in the views
+      firstShNode->Internal->EventsDisabled = false;
```

Comment:
It is always safer to store the old value of EventsDisabled and restore to the old value instead of hardcoding to false (the behavior is incorrect if a method that temporarily disables events calls another one that temporarily disables events, because this method re-enables events).
Name of the flag could be more specific, too, such as ItemUpdateEventsDisabled.